enterprise/wsproxy/keycache.go

```diff
@@ -0,0 +1,192 @@
+package wsproxy
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"golang.org/x/xerrors"
+
+	"cdr.dev/slog"
+
+	"github.com/coder/coder/v2/coderd/cryptokeys"
+	"github.com/coder/coder/v2/codersdk"
+	"github.com/coder/coder/v2/enterprise/wsproxy/wsproxysdk"
+	"github.com/coder/quartz"
+)
+
+var _ cryptokeys.Keycache = &CryptoKeyCache{}
+
+type CryptoKeyCache struct {
+	ctx    context.Context
+	cancel context.CancelFunc
+	client *wsproxysdk.Client
+	logger slog.Logger
+	Clock  quartz.Clock
+
+	keysMu sync.RWMutex
+	keys   map[int32]codersdk.CryptoKey
+	latest codersdk.CryptoKey
+	closed bool
+}
+
+func NewCryptoKeyCache(ctx context.Context, log slog.Logger, client *wsproxysdk.Client, opts ...func(*CryptoKeyCache)) (*CryptoKeyCache, error) {
+	cache := &CryptoKeyCache{
+		client: client,
+		logger: log,
+		Clock:  quartz.NewReal(),
+	}
+
+	for _, opt := range opts {
+		opt(cache)
+	}
+
+	m, latest, err := cache.fetch(ctx)
+	if err != nil {
+		return nil, xerrors.Errorf("initial fetch: %w", err)
+	}
+	cache.keys, cache.latest = m, latest
+	cache.ctx, cache.cancel = context.WithCancel(ctx)
+
+	go cache.refresh()
+
+	return cache, nil
+}
+
+func (k *CryptoKeyCache) Signing(ctx context.Context) (codersdk.CryptoKey, error) {
+	k.keysMu.RLock()
+
+	if k.closed {
+		k.keysMu.RUnlock()
+		return codersdk.CryptoKey{}, cryptokeys.ErrClosed
+	}
+
+	latest := k.latest
+	k.keysMu.RUnlock()
+
+	now := k.Clock.Now()
+	if latest.CanSign(now) {
+		return latest, nil
+	}
+
+	k.keysMu.Lock()
+	defer k.keysMu.Unlock()
+
+	if k.closed {
+		return codersdk.CryptoKey{}, cryptokeys.ErrClosed
+	}
+
+	if k.latest.CanSign(now) {
+		return k.latest, nil
+	}
+
+	var err error
+	k.keys, k.latest, err = k.fetch(ctx)
+	if err != nil {
+		return codersdk.CryptoKey{}, xerrors.Errorf("fetch: %w", err)
+	}
+
+	if !k.latest.CanSign(now) {
+		return codersdk.CryptoKey{}, cryptokeys.ErrKeyNotFound
+	}
+
+	return k.latest, nil
+}
```

Comment:
This looks very similar to the `dbcache` [`dbkeycache.go`](https://github.com/coder/coder/blob/0aa84b18a10382f70bda7ee4b1c3f4b6bdcd22f9/coderd/cryptokeys/dbkeycache.go#L108-L142).

Can we DRY up these keycache implementations?